### ISSUE TITLE
Collections' modelId method should be passed the Model as a second argument

### ISSUE THREAD

**Issue by amiller-gh** [jashkenas/backbone#3965]
Right now `modelId(attrs)` is only passed a plain-object representation of the model's attributes. This makes it difficult for application developers to create n-degree polymorphic Collection subclasses.

Contrived Example:

``` javascript
// A collection of mixed entity types that maintains sort order based 
// on their joinDate property

var Entities = Collection.extend({
  modelId(attrs){
    // What ID attribute to use? Not determined at runtime.
  },
  comparator: 'joinDate'
});


// These are only two entity types with different idAttributes, but assume 
// There may be up to (n) different entity types, each with different idAttributes, 
// determined by the database backing this example application.

var Company = Model.extend({
  idAttribute: 'companyId'
});

var Customer = Model.extend({
  idAttribute: 'customerId'
});

// Proceeding to add and remove Customers, Companies, etc, to the Entities
// Collection will mess up the `_byId` cache because the Entities collection
// is not aware of all the different possibilities for `idAttribute` at runtime and 
// a `modelId` method cannot be written to satisfy the needs of this app.

```

If the modelId callback took the form `modelId(attrs, model)`, where the second argument is the model in question, Backbone can retain backwards compatibility but allow application developers to do something like this:

``` javascript
// Always give precedence to the provided model's idAttribute. 
// Fall back to the Collection's idAttribute, and then to the default value `id`.

Entities.modelId = function modelId(attrs, model){
  return attrs[model.idAttribute || this.model.prototype.idAttribute || 'id'];
};
```

It may even make sense to make the above `modelId` function the default implementation and support polymorphic Collections out of the box. It still maintains backwards compatibility: For any implementation that is not already polymorphic, Collections will still function normally – models will have their idAttribute set when added to the Collection by the Collection's model constructor. Those that are already polymorphic, or have some type of non-standard idAttributes, will already have this modelId method overwritten.

Happy to throw up a PR if you think this is a good idea.

**Comment by jridgewell** [comment]
Please! But keep in mind, we've tried doing this [before](https://github.com/jashkenas/backbone/issues/3408#issuecomment-66223271) without much success.

**Comment by epicmiller** [comment]
PR is here: https://github.com/jashkenas/backbone/pull/3966

**Comment by prantlf** [comment]
We have used Backbone 1.1.2. Upgrade to 1.2.0 broke scenarios, where the "id" property exists in model's `attributes`, but it is not the collection's  `idAttribute`. It is not an uncommon scenario; imagine adding models to a "temporary" collection without `idAttribute`. No need for polymorfismus. For example:

```javascript
Version = Backbone.Model.extend({
  idAttribute: 'version_number'
});

version1 = new Version({
  version_number: 1,
  id: 2,
  name: 'First version'
});

version2 = new Version({
  version_number: 2,
  id: 2,
  name: 'Second version'
});

versions = new Backbone.Collection();
// Only the first version will be set to the collection
versions.reset([version1, version2]);
```

The referenced PR fixes this problem, but it always prefers the model's `idAttribute` to the collection's one. (`idAttribute` is set to "id" by default.) The collection's `idAttribute` will never be used.

#3133 broke behaviour of Backbone 1.1.2, #3966 breaks Backbone 1.2.0. @caseywebdev , is there a chance to support both scenarios? Preferable without changing the code base... I could think only about changing this PR to prefer `idAttribute` from the collection's model. It means creating collections, which do not know, which models will be added to them, without the `idAttribute` in their default model:

```javascript
TemporaryCollection = Backbone.Collection.extend({
  model: Backbone.Model.extend({
      idAttribute: null
  })
});

versions = new TemporaryCollection();
// collection's idAttribute is null, fallback to model's one lets both models set
versions.reset([version1, version2]);
```

Is there a better possibility for this scenario?

**Comment by jgonggrijp** [comment]
Closed in #3966. Related questions welcome in new tickets.
